### PR TITLE
Refactor Tile calling API to be more robust

### PR DESCRIPTION
**Pull request by minkezhang**
* Swap `TERRAIN_TYPE_BLOCKED` check for bounded cost check
* Refactor `tileastar.Path` to take in value-instance `MapCoordinate` instead of a reference to a `Tile` object

### CODE REVIEW

File: lib/hpf/graph.go [hunk 1]
```diff
@@ -114,6 +115,103 @@ func BuildGraph(tm *tile.Map, tileDimension *rtsspb.Coordinate) (*Graph, error)
...
+// AddEphemeralNode adds a temporary AbstractNode to the Graph and connects it
+// to the rest of the cluster via AbstractEdge instances.
+//
+// Function returns a UUID which needs to be tracked by the caller to be passed
+// into RemoveEphemeralNode. This function is a no-op if the input coordinates
+// match a non-ephemeral AbstractNode.
+//
+// TODO(minkezhang): Support rollback in case errors happen so that
+// AddEphemeralNode is idempotent.
+func AddEphemeralNode(tm *tile.Map, g *Graph, t utils.MapCoordinate) (int64, error) {
```
Comment (RyanzpLee) [inline lib/hpf/graph.go:127]: Okay, so if my understanding is correct, ephemeralnodes are short lasting - temporary nodes

File: lib/hpf/entrance.go [hunk 2]
```diff
@@ -279,7 +279,7 @@ func buildTransitionsAux(m *tile.Map, s1, s2 *rtsspb.CoordinateSlice) ([]*rtsspb
 			return nil, err
 		}
 
-		if (m.TileFromCoordinate(t1).TerrainType() != rtscpb.TerrainType_TERRAIN_TYPE_BLOCKED) && (m.TileFromCoordinate(t2).TerrainType() != rtscpb.TerrainType_TERRAIN_TYPE_BLOCKED) {
+		if m.C[m.TileFromCoordinate(t1).TerrainType()] < math.Inf(0) && m.C[m.TileFromCoordinate(t2).TerrainType()] < math.Inf(0) {
```
Comment (RyanzpLee) [inline lib/hpf/entrance.go:282]: Transition nodes  are for 3 tiles wide. Are these transition nodes just to allow us to jump from one cluster to the next, i.e connecting from graph to another?

**Review by RyanzpLee — DISMISSED** [review]

File: lib/hpf/entrance.go (hunk 2 above)
Comment (minkezhang) [reply lib/hpf/entrance.go:282]: A `Transition` is a pair of adjacent `AbstractNode` instances which straddle on the border between two clusters. We space the `Transition` instances a minimum of three tiles apart based on Botea. These objects are constructed only during `Graph` construction (and eventually while mutating `tile.Map`), and should not be a protobuf struct.

But to answer your question, yes, they are used to construct the inter-cluster `AbstractEdge` instances.

File: lib/hpf/graph.go (hunk 1 above)
Comment (minkezhang) [reply lib/hpf/graph.go]: Correct.

**Pull request merged**